packages/express/src/express.ts

```diff
@@ -1,46 +1,175 @@
-import { Container, ContainerAware, Plugin, Logger } from '@debugr/core';
-import { HttpFormatter } from '@debugr/http-formatter';
-import { Handler, ErrorRequestHandler } from 'express';
-import { NormalizedOptions, Options } from './types';
-import { normalizeOptions } from './utils';
-import { logHttpRequest } from './middleware/request';
-import { logHttpResponse } from './middleware/response';
+import { Plugin, Logger, LogLevel, TContextBase, TContextShape } from '@debugr/core';
+import { Handler, ErrorRequestHandler, Request, Response } from 'express';
+import { HttpLogEntry, NormalizedOptions, Options } from './types';
+import {
+  filterHeaders,
+  isCaptureEnabled,
+  normalizeContentLength,
+  normalizeContentType,
+  normalizeOptions,
+} from './utils';
 
-export class ExpressLogger implements ContainerAware, Plugin {
-  readonly id: string = 'express';
+export class ExpressLogger<
+  TTaskContext extends TContextBase = TContextBase,
+  TGlobalContext extends TContextShape = {},
+> implements Plugin<Partial<TTaskContext>, TGlobalContext>
+{
+  public readonly id: string = 'express';
+
+  public readonly entryFormat: 'http' = 'http';
 
   private readonly options: NormalizedOptions;
 
-  private logger: Logger;
+  private logger: Logger<Partial<TTaskContext>, TGlobalContext>;
 
-  constructor(options?: Options) {
+  public constructor(options?: Options) {
     this.options = normalizeOptions(options);
   }
 
-  injectContainer(container: Container): void {
-    this.logger = container.get('logger');
-
-    const pluginManager = container.get('pluginManager');
+  public static create<TTaskContext extends TContextBase, TGlobalContext extends TContextShape>(
+    options?: Options,
+  ): ExpressLogger<Partial<TTaskContext>, TGlobalContext> {
+    return new ExpressLogger<Partial<TTaskContext>, TGlobalContext>(options);
+  }
 
-    if (!pluginManager.has('http')) {
-      pluginManager.register(new HttpFormatter());
-    }
+  public injectLogger(logger: Logger<Partial<TTaskContext>, TGlobalContext>): void {
+    this.logger = logger;
   }
 
-  createRequestHandler(): Handler {
+  public createRequestHandler(): Handler {
     return (req, res, next) => {
-      this.logger.ensureFork(() => {
-        logHttpRequest(this.logger, this.options, req);
-        logHttpResponse(this.logger, this.options, res).finally(() => this.logger.flush());
+      this.logger.runTask(() => {
+        this.logHttpRequest(this.logger, this.options, req);
+        this.logHttpResponse(this.logger, this.options, res).finally(() => this.logger.flush());
         next();
       });
     };
   }
```

Comment:
skoro bych pro jistotu udělal něco jako:
```typescript
return async (req, res, next) => {
  return this.logger.runTask(async () => {
    this.logHttpRequest(this.logger, this.options, req);
    this.logHttpResponse(this.logger, this.options, res).finally(() => this.logger.flush());
    return next();
  });
};
```

čistě pro případ, že by `next()` náhodou vrátilo Promise - middleware v expressjs určitě může bejt asynchronní, jen musí na konci zavolat `next()`, a nevím, jak je to interně pošetřený v expressu..

Reply:
ok